### ISSUE TITLE
GraphTileHeader Bit Fields Differ on 32bit and 64bit Platforms

### ISSUE THREAD

**Issue by kevinkreiser** [valhalla/valhalla#1675]
There are some  typos in `GraphTileHeader`. Namely this section:

```c++
  // Currently there can only be twice as many transitions as there are nodes,
  // but in practice the number should be much less.
  uint32_t transitioncount_ : 22; // Number of node transitions
  uint32_t turnlane_count_ : 21;  // Number of turnlane records
  uint64_t transfercount_ : 16;   // Number of transit transfer records
  uint64_t spare2_ : 7;
```

I think what we were going for was squeezing all of this into a single 64bit value. Sadly, because we both mixed 32bit ints and overstepped the spare we run into problems. Namely, that the size of the structure on x64 systems is 272 where its only 268 on x86 systems. At any rate the code should look like this:

```c++
  // Currently there can only be twice as many transitions as there are nodes,
  // but in practice the number should be much less.
  uint64_t transitioncount_ : 22; // Number of node transitions
  uint64_t turnlane_count_ : 21;  // Number of turnlane records
  uint64_t transfercount_ : 16;   // Number of transit transfer records
  uint64_t spare2_ : 5;
```

However making the above change breakings the tile format because of the way the compiler spreads the bits out over the various bytes of the bit field when you cross fields over the word boundaries. This is sad because it means we need to go right to Valhalla v4 like a month after we hit v3...

I want to do 2 things:

- [ ] Add a 32bit docker image to CI so that we are always building the tests for at least one 32bit platform
- [ ] Modify the structure of `GraphTileHeader` so that the result is the same as the broken version today on 64bit systems but that 32bit systems also see the header structure this way. This would essentially just be a breakfix for 32bit systems (since most all tile data is created on 64bit systems)

**Comment by dnesbitt61** [comment]
Would this keep the layout the same on x64?

 uint32_t transitioncount_ : 22; // Number of node transitions
 uint32_t spare3_ : 10;

  uint32_t turnlane_count_ : 21;  // Number of turnlane records
  uint32_t spare4_ 11;

  uint64_t transfercount_ : 16;   // Number of transit transfer records
  uint64_t spare2_ : 48;

**Comment by kevinkreiser** [comment]
Ok so I'm working to see if i can do the modification suggested above. Currently the `actor` unit test fails on 32bit machines. What I've done to debug this is to create a gdb command file called `test.gdb`:

```
set width 0
set height 0
set verbose off

b search.cc:532
commands 1
  p "spare1_"
  p/d (void*)&tile->header_->spare1_ - (void*)&tile->header_->spare1_
  p "transitioncount_"
  p/d (void*)&tile->header_->transitioncount_ - (void*)&tile->header_->spare1_
  p "turnlane_count_"
  p/d (void*)&tile->header_->turnlane_count_ - (void*)&tile->header_->spare1_
  p "transfercount_"
  p/d (void*)&tile->header_->transfercount_ - (void*)&tile->header_->spare1_
  p "spare2_"
  p/d (void*)&tile->header_->spare2_ - (void*)&tile->header_->spare1_
  p "departurecount_"
  p/d (void*)&tile->header_->departurecount_ - (void*)&tile->header_->spare1_
  p "stopcount_"
  p/d (void*)&tile->header_->stopcount_ - (void*)&tile->header_->spare1_
  p "routecount_"
  p/d (void*)&tile->header_->routecount_ - (void*)&tile->header_->spare1_
  p "schedulecount_"
  p/d (void*)&tile->header_->schedulecount_ - (void*)&tile->header_->spare1_
  p "signcount_"
  p/d (void*)&tile->header_->signcount_ - (void*)&tile->header_->spare1_
  p "access_restriction_count_"
  p/d (void*)&tile->header_->access_restriction_count_ - (void*)&tile->header_->spare1_
  p "admincount_"
  p/d (void*)&tile->header_->admincount_ - (void*)&tile->header_->spare1_
end

run 1
```

You can then see what addresses each field of the bit field belong to and thereby see how the different systems are arraning the fields. Heres how to run it:

    gdb --command=test.gdb --args test/actor

The output on 64bit systems is as follows:

```
$1 = "spare1_"
$2 = 0
$3 = "transitioncount_"
$4 = 8
$5 = "turnlane_count_"
$6 = 12
$7 = "transfercount_"
$8 = 16
$9 = "spare2_"
$10 = 16
$11 = "departurecount_"
$12 = 16
$13 = "stopcount_"
$14 = 16
$15 = "routecount_"
$16 = 24
$17 = "schedulecount_"
$18 = 24
$19 = "signcount_"
$20 = 24
$21 = "access_restriction_count_"
$22 = 32
$23 = "admincount_"
$24 = 32
```

On 32bit machines it looks like:

```

```

**Comment by kevinkreiser** [comment]
@dnesbitt61 I havent tried that combo yet but i have tried quite a few combos at this point haha!

**Comment by dnesbitt61** [comment]
I have a working version for x64 that keeps the bit alignment. I became more complicated - as several structures after transfer_count_ and spare2_ needed to be reworked. Basically spare2_ has to remain at 7 bits, then you move members from successive bit fields up until the 64 bit bdry is broken, add spare to make that equal 64, then redo the next structure. It looks a bit uglier (5 new spare fields are needed).

  // Currently there can only be twice as many transitions as there are nodes,
  // but in practice the number should be much less.
  uint64_t transitioncount_ : 22; // Number of node transitions
  uint64_t spare3_ : 10;
  uint64_t turnlane_count_ : 21;  // Number of turnlane records
  uint64_t spare4_ : 11;

  // Number of transit records
  uint64_t transfercount_ : 16;
  uint64_t spare2_ : 7;
  uint64_t departurecount_ : 24;
  uint64_t stopcount_ : 16;
  uint64_t spare5_ : 1;

  // Number of transit records (continued), sign count
  uint64_t routecount_ : 12;
  uint64_t schedulecount_ : 12;
  uint64_t signcount_ : 24;
  uint64_t spare6_ : 16;

  // Additional record counts
  uint64_t access_restriction_count_ : 24; // Number of access restriction records
  uint64_t admincount_ : 16;               // Number of admin records
  uint64_t spare7_ : 24;

**Comment by dnesbitt61** [comment]
A bit of a pain, but this would keep bit compatibility for x64 so we don't need to update to v4. Would this then be compatible with x32?

**Comment by kevinkreiser** [comment]
@dnesbitt61 yeah i already did this as well. sadly there is more to fix as reachability is still not working in 32bit

**Comment by kevinkreiser** [comment]
Confirmed. `NodeInfo` is missing a single bit after the `mode_change_` field. I think just adding it in as a spare will fix it.
